File: Rust_rustc/src/agent/sarsa.rs
use crate::agent;
use crate::config::Config;
use crate::q_table::QTable;
use crate::{Action, Reward, State};
use rand;
use rand::prelude::*;
use std::error::Error;
use std::f64;

pub struct Sarsa {
    alpha: f64,
    gamma: f64,
    eps: f64,

    rng: ThreadRng,
}

impl Sarsa {
    pub fn new(config: &Config) -> Result<Sarsa, Box<dyn Error>> {
        let alpha = config.get("AGENT_ALPHA")?.parse()?;
        let gamma = config.get("AGENT_GAMMA")?.parse()?;
        let eps: f64 = config.get("AGENT_EPSILON")?.parse()?;
        let rng = rand::thread_rng();
        Ok(Sarsa {
            alpha,
            gamma,
            eps,
            rng,
        })
    }
}

impl agent::Agent for Sarsa {
    fn action(&mut self, q_table: &QTable, s: State) -> Action {
        if self.rng.gen::<f64>() < self.eps {
            self.rng.gen_range(0, q_table.action_size)
        } else {
            argmax(&q_table.table[s])
        }
    }

    fn learn(&self, q_table: &mut QTable, s1: State, a1: Action, r: Reward, s2: State, a2: Action) {
        let alpha = self.alpha;
        let gamma = self.gamma;

        q_table.table[s1][a1] = (1.0 - self.alpha) * q_table.table[s1][a1]
            + alpha * (r + gamma * q_table.table[s2][a2]);
    }

    fn fix(&mut self) {
        self.alpha = 0.;
        self.eps = 0.;
    }
}

fn argmax<T: PartialOrd>(slice: &[T]) -> usize {
    let mut res = 0;
    for i in 1..slice.len() {
        if slice[res] < slice[i] {
            res = i
        }
    }
    return res;
}

File: Ruby_MRI/agent/sarsa.rb
class Sarsa
  def initialize(config)
    @alpha = config.config["AGENT_ALPHA"].to_f
    @gamma = config.config["AGENT_GAMMA"].to_f
    @eps = config.config["AGENT_EPSILON"].to_f
  end

  def action(q_table, s)
    if rand < @eps
      return rand(0...q_table.action_size)
    end
    argmax(q_table.table[s])
  end

  def learn(q_table, s1, a1, r, s2, a2)
    q_table.table[s1][a1] = \
      (1.0 - @alpha) * q_table.table[s1][a1] + @alpha * (r + @gamma * q_table.table[s2][a2])
  end

  def fix
    @alpha = 0.0
    @eps = 0.0
  end

  def argmax(arr)
    res = 0
    (1...arr.length).each do |i|
      if arr[res] < arr[i]
        res = i
      end
    end
    res
  end
end
